### PR TITLE
[#36204] Add new GitHub plugin tab enabling users to copy git actions

### PR DESCRIPTION
**Pull request by tessi**
### Features

* It registers a new tab to the new core work package tab system.
* The GitHub tab is designed according to the figma prototype
* It allows copying default branch names, commit messages, and git-feature-branch-creation shell commands using [`tickety-tick-formatter`](https://github.com/bitcrowd/tickety-tick-formatter) under the hood which we just extracted from `tickety-tick` for this purpose.
* Two new icons are introduced to our icon font, a lighter console icon and a git-merge icon
* It is intentionally not yet displaying pull requests (this part will be continued in a separate PR)
* the ruby-side of the GitHub integration was refactored slightly to be without rubocop warnings (there are no intentional functional changes)

https://user-images.githubusercontent.com/206108/108431746-94d08180-7243-11eb-83b0-61bf13a4c6a6.mov

### State of this PR 🏗️ 

This PR is complete feature-wise but still has a couple of TODOs:

* [ ] get an angular-architecture review from @oliverguenther 
* [x] write js tests
* [x] write feature specs
* [x] get #8999 merged to unblock merging this PR
* [x] fix linter errors
* [x] https://github.com/bitcrowd/tickety-tick-formatter was created as a package for this PR. We should probably also review that to some degree
* [x] CLA

### CODE REVIEW

File: frontend/src/app/components/op-context-menu/op-context-menu-handler.ts [hunk 1]
```diff
@@ -18,11 +18,11 @@ export abstract class OpContextMenuHandler extends UntilDestroyedMixin {
    * Called when the service closes this context menu
    */
   public onClose() {
-    this.afterFocusOn.focus();
+    this.afterFocusOn.trigger('focus');
   }
 
   public onOpen(menu:JQuery) {
-    menu.find('.menu-item').first().focus();
+    menu.find('.menu-item').first().trigger('focus');
```
Comment (tessi) [inline frontend/src/app/components/op-context-menu/op-context-menu-handler.ts:25]: this is a drive-by fix - I noticed that `focus()` is a deprecated method while reading some menu handler code. The deprecation notice says to use `trigger()` instead.

File: modules/github_integration/frontend/module/tab-header/tab-header.template.html [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<div class="github-pr-header">
+  <h3 class="title">
+    <op-icon icon-classes="button--icon icon-merge-branch"></op-icon>
+    {{text.title}}
+  </h3>
+  <ul class="toolbar-items hide-when-print">
+    <li class="toolbar-item">
+      <button class="button github-create-pr"
+          [disabled]="true"
+          [attr.aria-label]="text.createPrButtonDescription">
```
Comment (b12f) [inline modules/github_integration/frontend/module/tab-header/tab-header.template.html]: It's usually good form to make buttons `type="button"`. They're `type="submit"` by default, and if this tab header gets included into a larger form switching tabs might submit the form by accident.

Comment (tessi) [reply modules/github_integration/frontend/module/tab-header/tab-header.template.html]: I made all buttons `type="button"`, thanks for the recommendation :)

**Comment by tessi** [comment]
Update to the video in the PR description: The copy button is now placed next to the textarea, but looks attached especially in the one-line version.


https://user-images.githubusercontent.com/206108/108552745-01a25500-72f2-11eb-9395-902d44bdaa2d.mov

File: frontend/src/app/components/datetime/timezone.service.spec.ts [hunk 3]
```diff
@@ -59,7 +59,7 @@ describe('TimezoneService', function () {
       ]
     });
 
-    timezoneService = TestBed.get(TimezoneService);
+    timezoneService = TestBed.inject(TimezoneService);
```
Comment (tessi) [inline frontend/src/app/components/datetime/timezone.service.spec.ts:62]: Drive-by fix 🚜: this (and following similar changes) are fixing a deprecation warning I spotted while reading code

File: frontend/src/app/modules/hal/resources/work-package-resource.spec.ts [hunk 4]
```diff
@@ -48,8 +48,6 @@ import {WorkPackageNotificationService} from 'core-app/modules/work_packages/not
 import {WorkPackagesActivityService} from 'core-components/wp-single-view-tabs/activity-panel/wp-activity.service';
 import {TimezoneService} from 'core-components/datetime/timezone.service';
 import {APIV3Service} from 'core-app/modules/apiv3/api-v3.service';
-import { Tab } from 'core-app/components/wp-single-view-tabs/tab/tab';
-import { TabComponent } from 'core-app/components/wp-single-view-tabs/tab/tab.component';
```
Comment (tessi) [inline frontend/src/app/modules/hal/resources/work-package-resource.spec.ts]: These were just unused and introduced in my former PR. I wonder why the linter didn't complain 🤔  Anyways, now they're gone

**Comment by lindenthal** [comment]
Hi @tessi,
can you please provide the [CLA](https://docs.openproject.org/development/#openproject-contributor-license-agreement-cla) for your contribution?
Thanks
Niels

**Comment by tessi** [comment]
Note: The "Create PR" Button was removed again as we decided this feature does not makes sense at the moment

**Comment by tessi** [comment]
> Hi @tessi,
> can you please provide the [CLA](https://docs.openproject.org/development/#openproject-contributor-license-agreement-cla) for your contribution?
> Thanks
> Niels

Note: We talked about this in a private conversation - from my perspective the CLA discussion is resolved ✅

File: .rubocop.yml [hunk 5]
```diff
@@ -7,6 +7,12 @@ AllCops:
   Exclude:
     - db/schema.rb
 
+
+Gemspec/RequiredRubyVersion:
+  Exclude:
+    - modules/**/*.gemspec
```
Comment (tessi) [inline .rubocop.yml:13]: IMO: inline module "gems" should not define their own ruby version and instead rely on the main definition

File: .rubocop.yml [hunk 6]
```diff
@@ -282,5 +288,12 @@ Style/HashTransformKeys:
...
-Rspec/MultipleMemoizedHelpers:
+
+RSpec/MultipleMemoizedHelpers:
+  Max: 15
+
+RSpec/MultipleExpectations:
   Max: 15
+
+RSpec/ExampleLength:
+  Max: 25
```
Comment (tessi) [inline .rubocop.yml:299]: In my attempt to bring you a github_integration without rubocop warnings, I stumbled over some RSpec-related rules which didn't have good defaults yet (in my opinion, of course).

I'd suggest these as relaxed good defaults as a starting point. We can probably adjust these to our needs over time.

**Review by aleixsuau — COMMENTED** [review]
Hi @tessi ,

The frontend part LGTM. Regarding the Ruby part, I think it would be better if someone else more suited than me could take a look.

Thanks!

File: frontend/src/global_styles/fonts/_openproject_icon_definitions.scss [hunk 7]
```diff
@@ -1,1536 +1,1548 @@
 @mixin icon-mixin-accessibility {
-    content: "\f101";
+    content: '\f101';
```
Comment (tessi) [inline frontend/src/global_styles/fonts/_openproject_icon_definitions.scss:2]: code climate wanted me to change just two of these definitions but had all these quotes in their internal list of "known bad formatting errors". So I fixed them all instead of just my two quotes. 🤷

File: modules/github_integration/lib/open_project/github_integration/patches/api/work_package_representer.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
+module OpenProject::GithubIntegration
+  module Patches
+    module API
+      module WorkPackageRepresenter
+        module_function
+
+        def extension
+          ->(*) do
+            link :github,
+                 uncacheable: true do
```
Comment (ulferts) [inline modules/github_integration/lib/open_project/github_integration/patches/api/work_package_representer.rb]: The part of caching in OP relevant here works by having most of the JSON response cached and then removing parts from it. As long as the part is valid for all users, as is the case here, we only need to make sure that those users not allowed to see it don't. For that, OP offers `cache_if`. For an example see the [ProjectRepresenter](https://github.com/opf/openproject/blob/dev/lib/api/v3/projects/project_representer.rb#L71).

You can thus simply move the condition into the `cache_if` lambda.

File: modules/github_integration/config/locales/js-en.yml [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+          message: Message
+          message_help: Copy the default commit message.
+          cmd: Command
+          cmd_help: Copy a shell command which creates a new branch and an empty commit using the default branch name and commit message.
+          title: Copy to Clipboard
+          copy_button_help: Copy to clipboard
+          copy_success: ✅ Copied!
+          copy_error: ❌ Copy failed!
+      tab_prs:
+        empty: There are no Pull Requests linked yet. Link an existing PR by using the code <code>OP#%{wp_id}</code> in the PR description or create a new PR.
```
Comment (ulferts) [inline modules/github_integration/config/locales/js-en.yml]: I assume "Pull Request" is not a name so it would need to be written with lower case -> "pull request"

File: modules/github_integration/lib/open_project/github_integration/engine.rb [hunk 10]
```diff
@@ -52,5 +54,16 @@ class Engine < ::Rails::Engine
       ::OpenProject::Notifications.subscribe('github.issue_comment',
                                              &NotificationHandlers.method(:issue_comment))
     end
+
+    initializer 'github.permissions' do
+      OpenProject::AccessControl.map do |ac_map|
+        ac_map.project_module(:github, {}) do |pm_map|
+          pm_map.permission(:show_github_content, {}, {})
+        end
```
Comment (ulferts) [inline modules/github_integration/lib/open_project/github_integration/engine.rb:62]: an I18n key for `project_module_github` is missing which is why it falls back to 'Github' instead of displaying 'GitHub':

image

Comment (ulferts) [reply modules/github_integration/lib/open_project/github_integration/engine.rb:62]: Same is true for `permission_show_github_content`

image

File: modules/github_integration/lib/open_project/github_integration/patches/api/work_package_representer.rb (hunk 8 above)
Comment (ulferts) [reply modules/github_integration/lib/open_project/github_integration/patches/api/work_package_representer.rb]: On second thoughts, it should be enough to reduce the check to only:

```
cache_if: -> { current_user.allowed_to?(:show_github_content, represented.project) }
```

The permission will only be granted if the project supports it, i.e. if the module is enabled.

**Review by ulferts — CHANGES_REQUESTED** [review]
@tessi, I only found some minor things. So should be good to merge soon.

File: modules/github_integration/lib/open_project/github_integration/patches/api/work_package_representer.rb (hunk 8 above)
Comment (tessi) [reply modules/github_integration/lib/open_project/github_integration/patches/api/work_package_representer.rb]: cool, wouldn't have guess that from the name `cache_if` (thought it sometimes caches and sometimes not). But great to know, makes the code much shorter

File: modules/github_integration/config/locales/js-en.yml (hunk 9 above)
Comment (tessi) [reply modules/github_integration/config/locales/js-en.yml]: true! I somehow thought they made it a trademark/name. but apparently not. fixed!

**Comment by tessi** [comment]
@ulferts thanks for the review 💚 (TIL `cache_if`)  all points are addressed now

File: .rubocop.yml (hunk 6 above)
Comment (ulferts) [reply .rubocop.yml:299]: :+1:

File: .rubocop.yml (hunk 5 above)
Comment (ulferts) [reply .rubocop.yml:13]: :+1:

**Review by oliverguenther — APPROVED** [review]
